File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        r = len(matrix)
        c = len(matrix[0])
        
        dp = [[0 for _ in range(c)] for _ in range(r)]
        dp[0] = matrix[0]
        
        for y in range(1, r):
            for x in range(c):
                if x == 0:
                    dp[y][x] = min(dp[y-1][x], dp[y-1][x + 1]) + matrix[y][x]
                elif x == c - 1:
                    dp[y][x] = min(dp[y - 1][x], dp[y - 1][x - 1]) + matrix[y][x]
                else:
                    dp[y][x] = min(dp[y - 1][x], dp[y - 1][x - 1], dp[y - 1][x + 1]) + matrix[y][x]
                    
        return min(dp[-1])
        

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.rs
use std::cmp;

impl Solution {
    pub fn min_falling_path_sum(matrix: Vec<Vec<i32>>) -> i32 {
        let r:usize = matrix.len();
        let c:usize = matrix[0].len();
        let mut dp: Vec<Vec<i32>> = vec![vec![0; c]; r];
        
        for i in 0..c {
            dp[0][i] = matrix[0][i];
        }
        
        for y in 1..r {
            for x in 0..c {
                if x == 0 {
                    dp[y][x] = cmp::min(dp[y - 1][x], dp[y - 1][x + 1]) + matrix[y][x];
                } else if x == c - 1 {
                    dp[y][x] = cmp::min(dp[y - 1][x], dp[y - 1][x - 1]) + matrix[y][x];
                } else {
                    dp[y][x] = cmp::min(dp[y - 1][x], cmp::min(dp[y - 1][x - 1], dp[y - 1][x + 1])) + matrix[y][x];
                }
            }
        }
        let mut answer = 10001;
        
        for i in 0..c {
            answer = cmp::min(answer, dp[r - 1][i]);
        }
        answer
    }
}